test/e2e/set/scyllacluster/scyllacluster_shardawareness.go

```diff
@@ -69,57 +66,131 @@ var _ = g.Describe("ScyllaCluster", func() {
 
-		framework.By("Waiting for the driver to establish connection to shards")
-		session, err := gocqlx.WrapSession(clusterConfig.CreateSession())
-		o.Expect(err).NotTo(o.HaveOccurred())
-		defer session.Close()
+		const (
+			// Ref: https://github.com/apache/cassandra/blob/cassandra-5.0/doc/native_protocol_v4.spec
+			optionsFrame   = `\x04\x00\x00\x00\x05\x00\x00\x00\x00`
+			scyllaShardKey = "SCYLLA_SHARD"
 
-		err = wait.PollUntilContextTimeout(ctx, 100*time.Millisecond, 5*time.Second, true, func(context.Context) (done bool, err error) {
-			return len(connections) == nrShards, nil
-		})
-		o.Expect(err).NotTo(o.HaveOccurred())
+			connectionAttempts = 10
+		)
 
-		shardAwareAttempts := 0
-		for sourcePort, addr := range connections {
-			// Control connection is also put in pool, and it always uses the default port.
-			if sourcePort == 0 {
-				o.Expect(addr).To(o.HaveSuffix(fmt.Sprintf("%d", nonShardAwarePort)))
-				continue
+		for shard := 0; shard < nrShards; shard++ {
+			port := shardPort(shard, nrShards)
+
+			for i := 0; i < connectionAttempts; i++ {
+				framework.By("Establishing connection number %d to shard number %d", i, shard)
+				stdout, stderr, err := utils.ExecWithOptions(f.ClientConfig(), f.KubeClient().CoreV1(), utils.ExecOptions{
+					Command:       []string{"/bin/sh", "-c", fmt.Sprintf(`echo -e '%s' | nc -p %d %s %d`, optionsFrame, port, hosts[0], shardAwarePort)},
+					Namespace:     clientPod.Namespace,
+					PodName:       clientPod.Name,
+					ContainerName: clientPod.Name,
+					CaptureStdout: true,
+					CaptureStderr: true,
+				})
+				o.Expect(err).NotTo(o.HaveOccurred(), stdout, stderr)
+				o.Expect(stderr).To(o.BeEmpty())
+				o.Expect(stdout).ToNot(o.BeEmpty())
+
+				fp := &frameParser{buf: bytes.NewBuffer([]byte(stdout))}
+				fp.SkipHeader()
+
+				scyllaSupported := fp.ReadStringMultiMap()
+				o.Expect(scyllaSupported[scyllaShardKey]).To(o.HaveLen(1))
+				o.Expect(scyllaSupported[scyllaShardKey][0]).To(o.Equal(fmt.Sprintf("%d", shard)))
 			}
-			o.Expect(addr).To(o.HaveSuffix(fmt.Sprintf("%d", shardAwarePort)))
-			shardAwareAttempts++
 		}
-
-		// Control connection used for shard number discovery, lands on some random shard.
-		// This connection is also put in pool, and driver only establish connections to missing shards
-		// using shard-aware-port.
-		// Connections to shard-aware-port are guaranteed to land on shard driver wants.
-		o.Expect(shardAwareAttempts).To(o.Equal(nrShards - 1))
 	})
 })
 
-type DialerFunc func(ctx context.Context, network, addr string) (net.Conn, error)
+func shardPort(shard, nrShards int) int {
+	const (
+		maxPort = 65535
+		minPort = 49152
+	)
+	maxRange := maxPort - nrShards + 1
+	minRange := minPort + nrShards - 1
+	r := rand.Intn(maxRange-minRange+1) + minRange
+	return r/nrShards*nrShards + shard
+}
+
+type frameParser struct {
```

Comment:
this might be a helper lib with a "sanity" unit test

Reply:
done